encode.go

```diff
@@ -1,163 +1,211 @@
 package bson
 
 import (
-	"bytes"
 	"fmt"
 	"io"
+	"math"
 	"reflect"
+	"sort"
 	"strconv"
 )
 
 // Encoder writes BSON values to an output stream.
 type Encoder struct {
 	w   io.Writer
-	buf *bytes.Buffer
+	buf []byte
 }
 
 // NewEncoder returns a new encoder that writes to w.
 func NewEncoder(w io.Writer) *Encoder {
 	return &Encoder{
 		w:   w,
-		buf: bytes.NewBuffer(make([]byte, 0, 512)),
+		buf: make([]byte, 0, 512),
 	}
 }
 
 // Encode writes the BSON encoding of v to the stream.
 func (enc *Encoder) Encode(v any) error {
-	enc.buf.Reset()
+	enc.buf = enc.buf[:0]
 	if err := enc.marshal(v); err != nil {
 		return fmt.Errorf("encode failed: %w", err)
 	}
-	_, err := enc.w.Write(enc.buf.Bytes())
+	_, err := enc.w.Write(enc.buf)
 	return err
 }
 
 func (enc *Encoder) marshal(v any) error {
-	switch v := v.(type) {
-	case Marshaler:
+	if v, ok := v.(Marshaler); ok {
 		raw, err := v.MarshalBSON()
 		if err != nil {
 			return err
 		}
-		enc.buf.Write(raw)
-	case A:
```

Comment:
Yeah, `A`, `D` and `M` are currently unsupported (to be precise only `D`).

They will be back in next PR when all of them will start implementing `bson.Marshaler`.